smartsim/experiment.py

```diff
@@ -159,28 +161,63 @@ def __init__(
 
-        self._control = Controller(launcher=self._launcher)
-
-        self.fs_identifiers: t.Set[str] = set()
+        self._fs_identifiers: t.Set[str] = set()
+        """Set of feature store identifiers currently in use by this
+        experiment"""
         self._telemetry_cfg = ExperimentTelemetryConfiguration()
+        """Switch to specify if telemetry data should be produced for this
+        experiment"""
+
+    def start_jobs(
+        self, job: Job, *jobs: Job, dispatcher: Dispatcher = DEFAULT_DISPATCHER
+    ) -> tuple[LaunchedJobID, ...]:
+        """Execute a collection of `Job` instances.
+
+        :param job: The job instance to start
+        :param jobs: A collection of other job instances to start
+        :param dispatcher: The dispatcher that should be used to determine how
+            to start a job based on its settings. If not specified it will
+            default to a dispatcher pre-configured by SmartSim.
+        :returns: A sequence of ids with order corresponding to the sequence of
+            jobs that can be used to query or alter the status of that
+            particular execution of the job.
+        """
 
-    def _set_dragon_server_path(self) -> None:
-        """Set path for dragon server through environment varialbes"""
-        if not "SMARTSIM_DRAGON_SERVER_PATH" in environ:
-            environ["SMARTSIM_DRAGON_SERVER_PATH_EXP"] = osp.join(
-                self.exp_path, CONFIG.dragon_default_subdir
-            )
+        def _start(job: Job) -> LaunchedJobID:
+            args = job.launch_settings.launch_args
+            env = job.launch_settings.env_vars
+            # >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+            # FIXME: Remove this cast after `SmartSimEntity` conforms to
+            #        protocol. For now, live with the "dangerous" type cast
+            # ---------------------------------------------------------------------
+            exe = t.cast("ExecutableProtocol", job.entity)
+            # <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+            dispatch = dispatcher.get_dispatch(args)
+            try:
+                launch_config = dispatch.configure_first_compatible_launcher(
+                    from_available_launchers=self._active_launchers,
+                    with_settings=args,
+                )
+            except errors.LauncherNotFoundError:
+                launch_config = dispatch.create_new_launcher_configuration(
+                    for_experiment=self, with_settings=args
+                )
+            # Save the underlying launcher instance. That way we do not need to
+            # spin up a launcher instance for each individual job, and it makes
+            # it easier to monitor job statuses
+            # pylint: disable-next=protected-access
+            self._active_launchers.add(launch_config._adapted_launcher)
+            return launch_config.start(exe, env)
```

Comment:
My brain might explode, but that is okay. So here, you're calling `.start` but where is `make_shell_format_fn` actually being ran? I see it set in each launch file? I know its being run?

Reply:
`make_shell_format_fn` is being called when `smartsim.settings.arguments.launch.alps`, `smartsim.settings.launch.mpi`, etc. are loaded by the interpreter and the function that is returned from that call is added into the default dispatcher's dispatch registry.

The dispatch registration then passes it (the `make_shell_format_fn` returned function) into a launcher adapter where it is finally called when the `LauncherAdapter.start` method is finally called (at that point it is at the `self._adapt` instance var)!